core/src/token_info/cached.rs

```diff
@@ -1,27 +1,35 @@
 use super::{TokenBaseInfo, TokenId, TokenInfoFetching};
 
-use anyhow::{Context as _, Result};
-use async_std::sync::RwLock;
+use anyhow::{anyhow, Context as _, Result};
+use async_std::sync::{Mutex, RwLock};
 use futures::{
-    future::{BoxFuture, FutureExt},
+    future::{BoxFuture, FutureExt, Shared},
     stream::{self, StreamExt as _},
 };
 use std::collections::HashMap;
 use std::sync::Arc;
 
-/**
- * Implementation of TokenInfoFetching that stores previously fetched information in an in-memory cache for fast retrieval.
- * TokenIds will always be fetched from the inner layer, as new tokens could be added at any time.
- */
+/// Implementation of TokenInfoFetching that stores previously fetched information in an in-memory cache for fast retrieval.
+/// TokenIds will always be fetched from the inner layer, as new tokens could be added at any time.
 pub struct TokenInfoCache {
-    cache: RwLock<HashMap<TokenId, TokenBaseInfo>>,
```

Comment:
I wonder if you could store something like MaybeDone<Shared<BoxFuture<TokenBaseInfo>>>, so there is only one hashmap. Then when trying to get, return result if ready or wait on shared future if pending.

Reply:
Yes we can. I was thinking about doing that but then decided not to until I get a feel for whether or not we want to include the change.